code/game/objects/structures/roof.dm

```diff
@@ -154,10 +161,27 @@
 	remove_under_roof(living)
 
 /datum/roof_master_node/proc/connect(location)
+	var/list/unprocesed_nodes = list()
 	for(var/obj/effect/roof_node/node in location)
-		node.link_master(src)
+		unprocesed_nodes += node
+
+
+	while(unprocesed_nodes.len)
+		var/obj/effect/roof_node/node = popleft(unprocesed_nodes)
+		var/list/new_nodes = node.link_master(src)
+		if(new_nodes)
+			unprocesed_nodes += new_nodes
+
+	var/list/unprocesed_roofs = list()
+
 	for(var/obj/structure/roof/roof in location)
-		roof.link_master(src)
+		unprocesed_roofs += roof
```

Comment:
Would be better to just figure out all the work that needs to be performed, and have that in a list. Likely better to add using |= so you don't add duplicates; but you may want to test what performs better w/ large use cases: as in is byond faster at checking for duplicates than just enqueueing them anyways and finding out we don't need to work on them when you come across a node already set up.